File: src/android/Gimbal2.java
package io.hpd.cordova.gimbal2;

/*
 * Cordova Plugin - Gimba v2
 * Denny Tsai <happydenn@happydenn.net>
 */

import java.util.Date;
import java.text.SimpleDateFormat;
import java.util.List;

import android.util.Log;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.gimbal.android.Gimbal;
import com.gimbal.android.BeaconManager;
import com.gimbal.android.BeaconEventListener;
import com.gimbal.android.BeaconSighting;
import com.gimbal.android.Beacon;
import com.gimbal.android.PlaceEventListener;
import com.gimbal.android.PlaceManager;
import com.gimbal.android.Visit;


public class Gimbal2 extends CordovaPlugin {

    private CallbackContext gimbalCallbackContext;
    private SimpleDateFormat simpleDateFormat;

    private BeaconManager beaconManager;
    private BeaconEventListener beaconEventListener;

    private PlaceManager placeManager;
    private PlaceEventListener placeEventListener;

	public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("initialize")) {
            if (args.length() != 1) return false;

            String apiKey = args.getString(0);
            Gimbal.setApiKey(cordova.getActivity().getApplication(), apiKey);

            if (gimbalCallbackContext == null) {
                gimbalCallbackContext = callbackContext;
            }

            if (simpleDateFormat == null) {
                simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
            }

            beaconEventListener = new BeaconEventListener() {
                @Override
                public void onBeaconSighting(BeaconSighting beaconSighting) {
                    Log.i("INFO", "onBeaconSighting: " + beaconSighting.toString());

                    Beacon beacon = beaconSighting.getBeacon();

                    JSONObject responseObject = new JSONObject();

                    try {
                        responseObject.put("event", "onBeaconSighting");
                        responseObject.put("RSSI", beaconSighting.getRSSI());
                        responseObject.put("datetime", simpleDateFormat.format(new Date(beaconSighting.getTimeInMillis())));
                        responseObject.put("beaconName", beacon.getName());
                        responseObject.put("beaconIdentifier", beacon.getIdentifier());
                        responseObject.put("beaconUniqueIdentifier", beacon.getUuid());
                        responseObject.put("beaconBatteryLevel", beacon.getBatteryLevel());
                        responseObject.put("beaconIconUrl", beacon.getIconURL());
                        responseObject.put("beaconTemperature", beacon.getTemperature());
                    } catch (JSONException e) {
                        Log.e("Error", e.getMessage(), e);
                    }

                    PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, responseObject);
                    pluginResult.setKeepCallback(true);

                    gimbalCallbackContext.sendPluginResult(pluginResult);
                }
            };

            beaconManager = new BeaconManager();
            beaconManager.addListener(beaconEventListener);



            placeEventListener = new PlaceEventListener() {

                @Override
                public void onVisitStart (Visit visit) {
                    Log.i("INFO", "onVisitStart: " + visit.toString());

                    JSONObject responseObject = new JSONObject();

                    try {
                        responseObject.put("event", "onBeginVisit");
                        responseObject.put("visitId", visit.getVisitID());
                        responseObject.put("placeId", visit.getPlace().getIdentifier());
                        responseObject.put("placeName", visit.getPlace().getName());
                        responseObject.put("placeAttributes", visit.getPlace().getAttributes());
                    } catch (JSONException e) {
                        Log.e("Error", e.getMessage(), e);
                    }

                    PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, responseObject);
                    pluginResult.setKeepCallback(true);

                    gimbalCallbackContext.sendPluginResult(pluginResult);
                }

                @Override
                public void onVisitEnd (Visit visit) {
                    Log.i("INFO", "onVisitStart: " + visit.toString());

                    JSONObject responseObject = new JSONObject();

                    try {
                        responseObject.put("event", "onEndVisit");
                        responseObject.put("visitId", visit.getVisitID());
                        responseObject.put("placeId", visit.getPlace().getIdentifier());
                        responseObject.put("placeName", visit.getPlace().getName());
                        responseObject.put("placeAttributes", visit.getPlace().getAttributes());
                    } catch (JSONException e) {
                        Log.e("Error", e.getMessage(), e);
                    }

                    PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, responseObject);
                    pluginResult.setKeepCallback(true);

                    gimbalCallbackContext.sendPluginResult(pluginResult);
                }

                @Override
                public void onBeaconSighting(BeaconSighting beaconSighting, List<Visit> visits) {
                    for (Visit visit: visits) {
                        Beacon beacon = beaconSighting.getBeacon();

                        JSONObject responseObject = new JSONObject();

                        try {
                            responseObject.put("event", "onBeaconSightingForVisit");
                            responseObject.put("visitId", visit.getVisitID());
                            responseObject.put("placeId", visit.getPlace().getIdentifier());
                            responseObject.put("RSSI", beaconSighting.getRSSI());
                            responseObject.put("datetime", simpleDateFormat.format(new Date(beaconSighting.getTimeInMillis())));
                            responseObject.put("beaconName", beacon.getName());
                            responseObject.put("beaconIdentifier", beacon.getIdentifier());
                            responseObject.put("beaconUniqueIdentifier", beacon.getUuid());
                            responseObject.put("beaconBatteryLevel", beacon.getBatteryLevel());
                            responseObject.put("beaconIconUrl", beacon.getIconURL());
                            responseObject.put("beaconTemperature", beacon.getTemperature());
                        } catch (JSONException e) {
                            Log.e("Error", e.getMessage(), e);
                        }

                        PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, responseObject);
                        pluginResult.setKeepCallback(true);

                        gimbalCallbackContext.sendPluginResult(pluginResult);

                    }
                }
            };

            placeManager = PlaceManager.getInstance();
            placeManager.addListener(placeEventListener);

            List<Visit> visits = placeManager.currentVisits();
            for (Visit visit: visits) {
                placeEventListener.onVisitStart(visit);
            }


            Gimbal.start();

            return true;
        }

        if (action.equals("startBeaconManager")) {
            if (beaconManager == null) return false;

            placeManager.startMonitoring();
            beaconManager.startListening();
            return true;
        }

        if (action.equals("stopBeaconManager")) {
            if (beaconManager == null) return false;

            placeManager.stopMonitoring();
            beaconManager.stopListening();
            return true;
        }

		return false;
	}
}

File: src/ios/Gimbal2.m
//
//  Gimbal2.m
//  cordova-plugin-gimbal2
//
//  Created by Denny Tsai on 4/21/15.
//
//

#import "Gimbal2.h"

@implementation Gimbal2 {
    NSString *statusCallbackId;
    NSDateFormatter *dateFormatter;
}

- (void)pluginInitialize {
    [super pluginInitialize];
    
    dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
    
    self.beaconManager = [[GMBLBeaconManager alloc] init];
    self.beaconManager.delegate = self;
    
    self.placeManager = [[GMBLPlaceManager alloc] init];
    self.placeManager.delegate = self;
}

- (void)initialize:(CDVInvokedUrlCommand *)command {
    NSString *apiKey = command.arguments[0];
    statusCallbackId = command.callbackId;
    
    [Gimbal setAPIKey:apiKey options:nil];
    
    [Gimbal start];
    
    NSArray * current = [self.placeManager currentVisits];
    for(GMBLVisit *visit in current) {
        [self placeManager:self.placeManager didBeginVisit:visit];
    }
}

- (void)startBeaconManager:(CDVInvokedUrlCommand *)command {
    [self.beaconManager startListening];
}

- (void)stopBeaconManager:(CDVInvokedUrlCommand *)command {
    [self.beaconManager stopListening];
}

- (void)beaconManager:(GMBLBeaconManager *)manager didReceiveBeaconSighting:(GMBLBeaconSighting *)sighting {
    GMBLBeacon *beacon = sighting.beacon;
    NSDictionary *returnData = @{
        @"event": @"onBeaconSighting",
        @"RSSI": [NSNumber numberWithInteger:sighting.RSSI],
        @"datetime": [dateFormatter stringFromDate:sighting.date],
        @"beaconName": (beacon.name != nil) ? beacon.name : [NSNull null],
        @"beaconIdentifier": (beacon.identifier != nil) ? beacon.identifier : [NSNull null],
        @"beaconUniqueIdentifier": (beacon.uuid != nil) ? beacon.uuid : [NSNull null],
        @"beaconBatteryLevel": [NSNumber numberWithFloat:beacon.batteryLevel],
        @"beaconIconUrl": (beacon.iconURL != nil) ? beacon.iconURL : [NSNull null],
        @"beaconTemperature": [NSNumber numberWithInteger:beacon.temperature],
    };
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:returnData];
    pluginResult.keepCallback = [NSNumber numberWithBool:YES];
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:statusCallbackId];
}

- (void) placeManager:(GMBLPlaceManager *)manager didBeginVisit:(GMBLVisit *)visit {
    GMBLPlace *place = visit.place;
    [self sendPlaceEvent:@"onBeginVisit" withVisitId:visit.visitID withPlace:place];
}

- (void)placeManager:(GMBLPlaceManager *)manager didEndVisit:(GMBLVisit *)visit {
    GMBLPlace *place = visit.place;
    [self sendPlaceEvent:@"onEndVisit" withVisitId:visit.visitID withPlace:place];
}

- (void) placeManager:(GMBLPlaceManager *)manager didReceiveBeaconSighting:(GMBLBeaconSighting *)sighting forVisits:(NSArray *)visits {
    for(GMBLVisit *visit in visits) {
        GMBLBeacon *beacon = sighting.beacon;
        NSDictionary *returnData = @{
                                     @"event": @"onBeaconSightingForVisit",
                                     @"visitId":visit.visitID,
                                     @"placeId":visit.place.identifier,
                                     @"RSSI": [NSNumber numberWithInteger:sighting.RSSI],
                                     @"datetime": [dateFormatter stringFromDate:sighting.date],
                                     @"beaconName": (beacon.name != nil) ? beacon.name : [NSNull null],
                                     @"beaconIdentifier": (beacon.identifier != nil) ? beacon.identifier : [NSNull null],
                                     @"beaconUniqueIdentifier": (beacon.uuid != nil) ? beacon.uuid : [NSNull null],
                                     @"beaconBatteryLevel": [NSNumber numberWithFloat:beacon.batteryLevel],
                                     @"beaconIconUrl": (beacon.iconURL != nil) ? beacon.iconURL : [NSNull null],
                                     @"beaconTemperature": [NSNumber numberWithInteger:beacon.temperature],
                                     };
        
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:returnData];
        pluginResult.keepCallback = [NSNumber numberWithBool:YES];
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:statusCallbackId];
    }
}

- (void) sendPlaceEvent:(NSString*) eventName withVisitId: (NSString*) visitId withPlace: (GMBLPlace *) place {
    NSMutableDictionary *theseAttributes = [[NSMutableDictionary alloc] init];
    NSDictionary *returnData = @{
                                 @"event":eventName,
                                 @"visitId": visitId,
                                 @"placeId":place.identifier,
                                 @"placeName":place.name,
                                 @"placeAttributes":theseAttributes
                                 };
    
    for (NSString *key in [place.attributes allKeys]) {
        NSString* value = [place.attributes stringForKey:key];
        [theseAttributes setObject:value forKey:key];
    }
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:returnData];
    pluginResult.keepCallback = [NSNumber numberWithBool:YES];
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:statusCallbackId];
}

@end
